distributed/tests/test_cancelled_state.py

```diff
@@ -533,7 +514,7 @@ async def release_all_futures():
 
     await lock_compute.release()
 
-    if not raise_error:
+    if not wait_for_processing and not raise_error:
```

Comment:
`test_resumed_cancelled_handle_compute` defeated me.
I could not make any sense of it.
I found it so horrifically complicated that it made me originally can this PR and restart from scratch with #6716, and then again with #6844.
This change makes it green again, but I am not sure that the tested stories are not highlighting any problems.

Reply:
I opened https://github.com/dask/distributed/pull/6905 with some explanations. I admit the test is a bit convoluted but I believe it is valuable

Reply:
Looks like the test is flaky now. I'll try investigating.

Reply:
My best guess for the flakyness is that the ordering with `wait_for_state(f3.key, "processing", s)` and `lock_compute.release` is very timing sensitive. Previously, the state machine was wired in such a way that this wouldn't matter but that is no longer the case

Reply:
You're right, if I add a sleep(1) just before `await lock_compute.release()`, the two tests with `wait_for_processing=False` fail deterministically.

It's a race condition of client.submit (client->scheduler comms) vs. distributed.Lock release (client->scheduler->worker).

Pushing a fix.